kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt

```diff
@@ -5,52 +5,253 @@ import io.kotest.matchers.Matcher
+   asList() should haveUpperBound(value, "ShortArray")
+   return this
+}
+
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun CharArray.shouldHaveUpperBound(value: Char): CharArray {
+   asList() should haveUpperBound(value, "CharArray")
+   return this
+}
+
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun IntArray.shouldHaveUpperBound(value: Int): IntArray {
+   asList() should haveUpperBound(value, "IntArray")
+   return this
+}
+
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun LongArray.shouldHaveUpperBound(value: Long): LongArray {
+   asList() should haveUpperBound(value, "LongArray")
    return this
 }
 
-infix fun <T : Comparable<T>> Array<T>.shouldHaveUpperBound(t: T): Array<T> {
-   asList().shouldHaveUpperBound(t)
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun FloatArray.shouldHaveUpperBound(value: Float): FloatArray {
+   asList() should haveUpperBound(value, "FloatArray")
    return this
 }
 
-infix fun <T : Comparable<T>, C : Collection<T>> C.shouldHaveUpperBound(t: T): C {
-   this should haveUpperBound(t)
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun DoubleArray.shouldHaveUpperBound(value: Double): DoubleArray {
+   asList() should haveUpperBound(value, "DoubleArray")
    return this
 }
 
-fun <T : Comparable<T>, C : Collection<T>> haveUpperBound(t: T) = object : Matcher<C> {
-   override fun test(value: C): MatcherResult {
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun <T : Comparable<T>> Array<T>.shouldHaveUpperBound(value: T): Array<T> {
+   asList() should haveUpperBound(value, "Array")
+   return this
+}
+
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun <T : Comparable<T>, C : Collection<T>> C.shouldHaveUpperBound(value: T): C {
+   this should haveUpperBound(value, null)
+   return this
+}
+
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun <T : Comparable<T>, I : Iterable<T>> I.shouldHaveUpperBound(value: T): I {
+   this should haveUpperBound(value, null)
+   return this
+}
+
+/**
+ * Verifies that all elements are less than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+fun <T : Comparable<T>, C : Collection<T>> haveUpperBound(value: T): Matcher<C> = haveUpperBound(value, null)
+
+private fun <T : Comparable<T>, I : Iterable<T>> haveUpperBound(t: T, name: String?): Matcher<I> = object : Matcher<I> {
+   override fun test(value: I): MatcherResult {
+      val name = name ?: value.containerName()
```

Comment:
As we are messing with this code already, perhaps we should guard against empty lists.

Empty lists will throw an error if they fail and pass silently otherwise.

We should place something like 'require(value.isNotEmpty())' or something similar. I think we do this in a few other collection matchers

Reply:
This is up for debate IMO. If you word `hasUpperBound(T)` as "all elements are less than or equal to T", then an empty set does evaluate to true, because there are no elements to violate the constraint. This is known as a [_vacuous truth_](https://en.wikipedia.org/wiki/Vacuous_truth) -- statement is true, but doesn't really say anything.

I argue current behavior is correct. If you want to check for presence of elements, there is `shouldNotBeEmpty` for that.

See also: [stdlib function `all`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/all.html)

@AlexCue987 @sksamuel What do you think?

Reply:
> Empty lists will throw an error if they fail and pass silently otherwise.

I apologize, I don't understand this sentence. Could you please reword it?

Reply:
> I argue current behavior is correct. If you want to check for presence of elements, there is shouldNotBeEmpty for that.

Exactly. If we want to check for empty collections it should happen on `shouldNotBeEmpty`, and not accidentally on `shouldHaveUpperBound`. Therefore, if you're making a check against a collection but it's actually empty, we should prevent such thing to happen

> statement is true, but doesn't really say anything.

So the test would basically do nothing? That's not a very useful test, is it?




https://github.com/kotest/kotest/blame/1360aa5a85e5e44bdf85ef04718ad24019ca47bf/kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/ranges/bein.kt#L54

https://github.com/kotest/kotest/blob/1360aa5a85e5e44bdf85ef04718ad24019ca47bf/kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/matchers.kt#L133

These are examples in our codebase when we do this kind of guard clause.


> See also: [stdlib function all](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/all.html)

The stdlib function is for general purpose, and not testing/assertion purposes. I think this comparison doesn't apply

Reply:
I don't have a horse in this race. There's just a decision to be made about which behavior is deemed more in line with the principle of least surprise. 
I argued that I'd like to follow mathematical logic - but I'll happily throw an error on empty container. It just must be consistent everywhere. That's the important bit.

> So the test would basically do nothing? That's not a very useful test, is it?

No, it doesn't "do nothing". It verifies that the statement is true, which it is. An assertion is just a logical statement. Now, is it desirable for `hasLowerBound` to be just `hasLowerBound`, or should it be a conjunction of `isNotEmpty && hasLowerBound`?

Reply:
It is also important to note that this PR doesn't modify previous behavior.

If all collection assertions should suddenly start rejecting empty containers, that's a separate issue for a separate PR. I'll open a ticket for it.

Reply:
> It is also important to note that this PR doesn't modify previous behavior. It would be a behavioral change if this proposed isNotEmpty condition was added.

Fair point, I'll resolve this as we can track it in the new ticket you created